packages/google_sign_in/google_sign_in_android/android/src/main/java/io/flutter/plugins/googlesignin/GoogleSignInPlugin.java

```diff
@@ -550,31 +727,67 @@ private String errorCodeForStatus(int statusCode) {
       }
     }
 
-    private void finishWithSuccess(Object data) {
-      pendingOperation.result.success(data);
+    private void finishWithSuccess() {
+      Objects.requireNonNull(pendingOperation.voidResult).success(null);
+      pendingOperation = null;
+    }
+
+    private void finishWithBoolean(Boolean value) {
+      Objects.requireNonNull(pendingOperation.boolResult).success(value);
+      pendingOperation = null;
+    }
+
+    private void finishWithUserData(Messages.UserData data) {
+      Objects.requireNonNull(pendingOperation.userDataResult).success(data);
       pendingOperation = null;
     }
 
     private void finishWithError(String errorCode, String errorMessage) {
-      pendingOperation.result.error(errorCode, errorMessage, null);
+      Messages.Result<?> result;
+      if (pendingOperation.userDataResult != null) {
+        result = pendingOperation.userDataResult;
+      } else if (pendingOperation.boolResult != null) {
+        result = pendingOperation.boolResult;
+      } else if (pendingOperation.stringResult != null) {
+        result = pendingOperation.stringResult;
+      } else {
+        result = pendingOperation.voidResult;
+      }
+      Objects.requireNonNull(result).error(new FlutterError(errorCode, errorMessage, null));
       pendingOperation = null;
     }
 
     private static class PendingOperation {
-      final String method;
-      final Result result;
-      final Object data;
-
-      PendingOperation(String method, Result result, Object data) {
+      final @NonNull String method;
+      final @Nullable Messages.Result<Messages.UserData> userDataResult;
+      final @Nullable Messages.Result<Void> voidResult;
+      final @Nullable Messages.Result<Boolean> boolResult;
+      final @Nullable Messages.Result<String> stringResult;
```

Comment:
This is a bit clunky, but it allowed introducing the type safe responses without major structural changes to the plugin. (And in at least some cases storing them is unavoidable due to the way the intent callbacks work.)